File: benchmark/languages/go/work/day08/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type point struct{ x, y, z int }

type edge struct {
	d    int64
	a, b int
}

var parent, rank_ []int

func find(x int) int {
	for parent[x] != x {
		parent[x] = parent[parent[x]]
		x = parent[x]
	}
	return x
}

func union(a, b int) bool {
	ra, rb := find(a), find(b)
	if ra == rb {
		return false
	}
	if rank_[ra] < rank_[rb] {
		ra, rb = rb, ra
	}
	parent[rb] = ra
	if rank_[ra] == rank_[rb] {
		rank_[ra]++
	}
	return true
}

func main() {
	part := 0
	if len(os.Args) > 1 {
		part, _ = strconv.Atoi(os.Args[1])
	}

	data, _ := os.ReadFile("input.txt")
	lines := strings.Split(strings.TrimSpace(string(data)), "\n")
	var pts []point
	for _, ln := range lines {
		ln = strings.TrimSpace(ln)
		if ln == "" {
			continue
		}
		f := strings.Split(ln, ",")
		x, _ := strconv.Atoi(strings.TrimSpace(f[0]))
		y, _ := strconv.Atoi(strings.TrimSpace(f[1]))
		z, _ := strconv.Atoi(strings.TrimSpace(f[2]))
		pts = append(pts, point{x, y, z})
	}

	n := len(pts)
	var edges []edge
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			dx := int64(pts[i].x - pts[j].x)
			dy := int64(pts[i].y - pts[j].y)
			dz := int64(pts[i].z - pts[j].z)
			d := dx*dx + dy*dy + dz*dz
			edges = append(edges, edge{d, i, j})
		}
	}
	sort.Slice(edges, func(i, j int) bool {
		return edges[i].d < edges[j].d
	})

	w := bufio.NewWriter(os.Stdout)
	defer w.Flush()

	if part == 0 || part == 1 {
		parent = make([]int, n)
		rank_ = make([]int, n)
		for i := range parent {
			parent[i] = i
		}
		cnt := 0
		for _, e := range edges {
			if cnt >= 1000 {
				break
			}
			union(e.a, e.b)
			cnt++
		}
		sizes := map[int]int{}
		for i := 0; i < n; i++ {
			sizes[find(i)]++
		}
		var sz []int
		for _, v := range sizes {
			sz = append(sz, v)
		}
		sort.Sort(sort.Reverse(sort.IntSlice(sz)))
		res := 1
		for i := 0; i < 3 && i < len(sz); i++ {
			res *= sz[i]
		}
		fmt.Fprintf(w, "Part 1: %d\n", res)
	}

	if part == 0 || part == 2 {
		parent = make([]int, n)
		rank_ = make([]int, n)
		for i := range parent {
			parent[i] = i
		}
		comps := n
		var lastA, lastB int
		for _, e := range edges {
			if union(e.a, e.b) {
				comps--
				if comps == 1 {
					lastA, lastB = e.a, e.b
					break
				}
			}
		}
		res := pts[lastA].x * pts[lastB].x
		fmt.Fprintf(w, "Part 2: %d\n", res)
	}
}

File: benchmark/languages/c/work/day08/main.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

typedef struct {
    long long x, y, z;
} Point;

typedef struct {
    double dist;
    int a, b;
} Edge;

// Union-Find
static int *parent;
static int *rank_arr;
static int *sz;

int find(int x) {
    while (parent[x] != x) {
        parent[x] = parent[parent[x]]; // path compression (halving)
        x = parent[x];
    }
    return x;
}

// Returns 1 if merged (were different), 0 if same
int unite(int a, int b) {
    a = find(a);
    b = find(b);
    if (a == b) return 0;
    if (rank_arr[a] < rank_arr[b]) { int t = a; a = b; b = t; }
    parent[b] = a;
    sz[a] += sz[b];
    if (rank_arr[a] == rank_arr[b]) rank_arr[a]++;
    return 1;
}

int cmp_edge(const void *p1, const void *p2) {
    const Edge *e1 = (const Edge *)p1;
    const Edge *e2 = (const Edge *)p2;
    if (e1->dist < e2->dist) return -1;
    if (e1->dist > e2->dist) return 1;
    return 0;
}

int cmp_int_desc(const void *a, const void *b) {
    return *(int*)b - *(int*)a;
}

int main(int argc, char *argv[]) {
    int part = 0;
    if (argc >= 2) part = atoi(argv[1]);

    FILE *f = fopen("input.txt", "r");
    if (!f) { fprintf(stderr, "Cannot open input.txt\n"); return 1; }

    Point *pts = NULL;
    int n = 0, cap = 0;

    char line[256];
    while (fgets(line, sizeof(line), f)) {
        // trim
        int len = (int)strlen(line);
        while (len > 0 && (line[len-1] == '\n' || line[len-1] == '\r' || line[len-1] == ' '))
            line[--len] = '\0';
        if (len == 0) continue;

        if (n == cap) {
            cap = cap ? cap * 2 : 64;
            pts = realloc(pts, cap * sizeof(Point));
        }
        sscanf(line, "%lld,%lld,%lld", &pts[n].x, &pts[n].y, &pts[n].z);
        n++;
    }
    fclose(f);

    // Build all edges
    long long num_edges = (long long)n * (n - 1) / 2;
    Edge *edges = malloc(num_edges * sizeof(Edge));
    long long ei = 0;
    for (int i = 0; i < n; i++) {
        for (int j = i + 1; j < n; j++) {
            double dx = (double)(pts[i].x - pts[j].x);
            double dy = (double)(pts[i].y - pts[j].y);
            double dz = (double)(pts[i].z - pts[j].z);
            edges[ei].dist = sqrt(dx*dx + dy*dy + dz*dz);
            edges[ei].a = i;
            edges[ei].b = j;
            ei++;
        }
    }

    // Sort edges by distance
    qsort(edges, num_edges, sizeof(Edge), cmp_edge);

    // Initialize Union-Find
    parent = malloc(n * sizeof(int));
    rank_arr = calloc(n, sizeof(int));
    sz = malloc(n * sizeof(int));
    for (int i = 0; i < n; i++) {
        parent[i] = i;
        sz[i] = 1;
    }

    long long part1_ans = 0;
    long long part2_ans = 0;

    // Process edges for Part 1: first 1000 connections (edges processed, regardless of merge)
    // Wait: re-read problem. "Connect together the 1000 pairs of junction boxes which are closest together"
    // The problem says "connect the 1000 pairs", i.e., process the 1000 shortest edges (even if same component).
    // But the example says "making the ten shortest connections" — and the 4th pair (431,825,988 and 425,690,689)
    // was already in same circuit but "nothing happens" — so it still counts as one of the ten!
    // So we process the 1000 shortest EDGES (pairs), not 1000 merges.

    int connections_made = 0;    // count of edges processed (regardless of merge)
    int num_components = n;
    int part2_done = 0;

    for (long long i = 0; i < num_edges; i++) {
        if (connections_made < 1000 || !part2_done) {
            // process this edge
            int merged = unite(edges[i].a, edges[i].b);
            if (merged) {
                num_components--;
                if (num_components == 1 && !part2_done) {
                    // Part 2: product of X coordinates
                    part2_ans = pts[edges[i].a].x * pts[edges[i].b].x;
                    part2_done = 1;
                }
            }
            connections_made++;

            if (connections_made == 1000 && !part1_ans) {
                // Compute Part 1 answer: product of 3 largest component sizes
                // Collect all component sizes
                int *sizes = malloc(n * sizeof(int));
                int sc = 0;
                for (int j = 0; j < n; j++) {
                    if (find(j) == j) {
                        sizes[sc++] = sz[j];
                    }
                }
                qsort(sizes, sc, sizeof(int), cmp_int_desc);
                part1_ans = (long long)sizes[0] * sizes[1] * sizes[2];
                free(sizes);
            }

            if (connections_made >= 1000 && part2_done) break;
        }
    }

    // If part2 not done yet (all merged before processing all needed edges), handle
    // (shouldn't happen in normal cases)

    if (part == 1 || part == 0) {
        printf("Part 1: %lld\n", part1_ans);
    }
    if (part == 2 || part == 0) {
        printf("Part 2: %lld\n", part2_ans);
    }

    free(pts);
    free(edges);
    free(parent);
    free(rank_arr);
    free(sz);

    return 0;
}
